projects/cdk/directives/auto-focus/autofocus.directive.ts

```diff
@@ -32,57 +32,79 @@ export class TuiAutoFocusDirective implements AfterViewInit {
+        @Inject(ElementRef)
+        private readonly elementRef: ElementRef<HTMLElement>,
+        @Inject(Renderer2)
+        private readonly renderer: Renderer2,
     ) {}
 
     ngAfterViewInit() {
         if (!this.autoFocus) {
             return;
         }
 
-        const element =
-            this.tuiFocusableComponent?.nativeFocusableElement ||
-            this.elementRef.nativeElement;
-
-        if (this.isIos) {
-            veryVerySadIosFix(this.injector);
+        if (this.isIos && this.isTextFieldElement()) {
+            this.ngZone.runOutsideAngular(() => this.iosWebkitAutofocus());
+        } else {
+            race(
+                timer(TIMEOUT),
+                this.animationFrame$.pipe(
+                    throttleTime(POLLING_TIME),
+                    map(() => this.element.closest(NG_ANIMATION_SELECTOR)),
+                    skipWhile(Boolean),
+                    take(1),
+                ),
+            ).subscribe(() => setNativeFocused(this.element));
         }
+    }
 
-        race(
-            timer(TIMEOUT),
-            this.animationFrame$.pipe(
-                throttleTime(POLLING_TIME),
-                map(() => element.closest(NG_ANIMATION_SELECTOR)),
-                filter(v => !v),
-                take(1),
-            ),
-        ).subscribe(() => {
-            setNativeFocused(element);
-        });
+    private get element(): TuiNativeFocusableElement {
+        return (
+            this.tuiFocusableComponent?.nativeFocusableElement ||
+            this.elementRef.nativeElement
+        );
     }
-}
 
-function veryVerySadIosFix(injector: Injector) {
-    const renderer = injector.get(Renderer2);
-    const {nativeElement} = injector.get(ViewContainerRef).element;
-    const decoy: HTMLElement = renderer.createElement('input');
+    private isTextFieldElement(): boolean {
+        return this.element.matches('input, textarea');
+    }
+
+    private iosWebkitAutofocus(): void {
+        const fakeInput: HTMLElement = this.renderer.createElement('input');
+
+        fakeInput.style.position = 'absolute';
+        fakeInput.style.opacity = '0';
+        fakeInput.style.height = '0';
 
-    decoy.style.position = 'absolute';
-    decoy.style.opacity = '0';
-    decoy.style.height = '0';
+        /**
+         * @note: ping-pong eager strategy hack
+         * After creating an element and bringing it into DOM,
+         * the browser automatically focuses on the invisible element.
+         * And then, after focus is triggered, we try to focus on target element, and if we managed to refocus,
+         * then we try to focus again on an invisible element, so that the keyboard slowly appears.
+         * This ping pong allows the keyboard to not overlap the modal window.
+         */
+        fakeInput.addEventListener('blur', () => setNativeFocused(fakeInput), {
+            once: true,
+        });
+
+        fakeInput.addEventListener('focus', () =>
```

Comment:
shouldn't we remove event listeners on destroy?

Reply:
fixed